include/envoy/http/async_client.h

```diff
@@ -75,6 +75,13 @@ class AsyncClient {
      */
     virtual void onTrailers(HeaderMapPtr&& trailers) PURE;
 
+    /**
+     * Called when both the local and remote have gracefully closed the stream.
+     * Useful for asymmetric cases where end_stream may not be bidirectionally observable.
+     * Note this is NOT called on stream reset.
+     */
+    virtual void onClosure() PURE;
```

Comment:
On the platform APIs, we use `onClose`. Can we replicate that here for consistency?

Reply:
All the others are `onNoun`, so I was just shooting for consistency.

Elsewhere in Envoy `onCompletion` is used, which is another possibility, but I wanted to better distinguish from `onReset`.

I don't have strong feelings about it though.

Reply:
personally I'd prefer `onCompletion` to `onClosure`, primarily to reduce confusion with the Swift concept of a "closure"/lambda. thoughts?

Reply:
I think the other thing I like slightly better about `onClosure` (and this is super subjective) is that "close" being an action makes it maybe sound like _either_ the remote or local terminals closing their end will trigger this, when in actuality this only fires when the stream enters a state of full closure.

Reply:
I don't feel _that_ strongly, so this is fine with me

Reply:
I would probably call this `onComplete()` which is how we name this elsewhere (and how it's named internally in the code you are refactoring).

Reply:
`onComplete` sounds good to me, since there's quite a bit of precedent in Envoy. Will go with that.